llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp

```diff
@@ -2617,8 +2621,40 @@ static VPRecipeBase *optimizeMaskToEVL(VPValue *HeaderMask,
   return nullptr;
 }
 
-/// Replace recipes with their EVL variants.
-static void transformRecipestoEVLRecipes(VPlan &Plan, VPValue &EVL) {
+void VPlanTransforms::optimizeMasksToEVL(VPlan &Plan) {
+  // Find the EVL-based header mask if it exists: icmp ult step-vector, EVL
+  VPInstruction *HeaderMask = nullptr;
+  for (VPRecipeBase &R : *Plan.getVectorLoopRegion()->getEntryBasicBlock()) {
+    if (match(&R, m_ICmp(m_VPInstruction<VPInstruction::StepVector>(),
+                         m_EVL(m_VPValue())))) {
+      HeaderMask = cast<VPInstruction>(&R);
+      break;
+    }
+  }
+  if (!HeaderMask)
+    return;
+
+  VPValue *EVL = HeaderMask->getOperand(1);
```

Comment:
Do you have plan to create helper function like `static VPValue *findEVLMask(VPlan &Plan)`?
If not, could we do 
```suggestion
  // Find the EVL-based header mask if it exists: icmp ult step-vector, EVL
  VPInstruction *HeaderMask = nullptr;
  VPValue *EVL;
  for (VPRecipeBase &R : *Plan.getVectorLoopRegion()->getEntryBasicBlock()) {
    if (match(&R, m_ICmp(m_VPInstruction<VPInstruction::StepVector>(),
                         m_EVL(m_VPValue(EVL))))) {
      HeaderMask = cast<VPInstruction>(&R);
      break;
    }
  }
  if (!HeaderMask)
    return;
```
?

Reply:
BTW, do we need to update vputils::isHeaderMask?